File: python/mgBin.py
""" mgBin.py (since written in a functional style, annotated for mypy) """
import fset
from listfs import * # for partition, unzip

def mrg(lst: list) -> frozenset:
  """ merge """
  return fset.fromList(lst)

def ck(labels: list) -> list: 
  """ remove first features from already matched labels[0] and labels[1] """
  [(nns,nps), ((),pps)] = labels
  return [(nns[1:],nps), ((),pps[1:])]

def t(lsos:list) -> list:
  """ remove LSOs with no features in their label """
  return [x for x in lsos if x[1][1]]

def soSize(so) -> int:
  """ calculate the size of a syntactic object """
  if isinstance(so,tuple) and len(so) == 2 and \
     isinstance(so[0],tuple) and ( len(so[0]) == 0 or isinstance(so[0][0],str) ): # lex
    return 1
  elif isinstance(so,frozenset): # fset
    return 1 + sum(map(soSize, fset.toList(so)))
  else:  # phtree
    return 1 + sum(map(soSize, so))

def maxx(lsos: list) -> tuple:
  """ given LSOs (of a WS), return LSO with largest SO (the head) """
  sofar = 0
  maxLSO = ((()),((),()))
  for lso in lsos:
    lsoSize = soSize(lso[0])
    if lsoSize > sofar:
      sofar = lsoSize
      maxLSO = lso
  return maxLSO

def smc(lsos: list) -> frozenset:
  """ if LSOs have no pos feature in common, return them """
  firstPosFeatures = [lso[1][1][0] for lso in lsos if lso[1][0] == ()]
  if len(firstPosFeatures) < len(set(firstPosFeatures)):
    return fset.fromList([])
  else:
    return fset.fromList(lsos)

def match(wss:list) -> tuple:
  """ given list of workspaces, return ([head,complement],otherLSOs) """
  (ws0,wss0) = (wss[0],wss[1:])
  (heads, others) = partition(lambda lso: lso[1][0] != (), fset.toList(ws0))
  if not(len(heads)==1):
    raise RuntimeError('match: 0 or >1 neg lsos in ws0')
  h = heads[0]
  f = h[1][0][0]
  (ics,iothers) = partition(lambda lso: lso[1][1][0] == f, others)
  if ics and wss0 == []:    # im
    return ([h,ics[0]], iothers)
  elif ics==[] and len(wss0)==1:   # em
    ws1 = wss0[0]
    lsos = fset.toList(ws1)
    c = maxx(lsos)
    if c[1][0] == () and c[1][1][0] == f:
      others1 = [x for x in lsos if x != c]
      return ([h,c], others + others1)
    else: RuntimeError('match: complement')
  else:
    raise RuntimeError('match')

def d(wss:list) -> frozenset:
  """ the derivational step: given list of workspaces, return derived workspace """
  (matched, movers) = match(wss)
  (sos,labels) = unzip(matched)
  return smc( t( list(zip( [mrg(sos)]+sos[1:], ck(labels))) + movers))

File: haskell/MgBin.hs
module MgBin where
import Data.Set (Set)
import qualified Data.Set as Set
import Data.List
import qualified Data.List as List

data Ft = C | D | N | V | A | P | Wh deriving (Show, Eq, Ord)
type Label = ([Ft], [Ft])
type Lex = ([String], Label)
data PhTree = Pl Lex | Ps [PhTree] deriving (Show, Eq, Ord)
data SO = L Lex | S (Set (SO)) | O PhTree deriving (Show, Eq, Ord)
type LSO = (SO, Label)
type WS = Set (LSO)

-- merge
mrg :: [SO] -> SO
mrg sos = S (Set.fromList sos)

-- already matched features can be `forgotten'
ck :: [Label] -> [Label]
ck [ (_:nns,nps), ([],_:pps) ] = [ (nns,nps), ([],pps) ]

-- constituents with no remembered features can be `forgotten'
t :: [LSO] -> [LSO]
t = filter (\lso -> snd lso /= ([],[]))

-- return number of nodes in SO (here we can disregard multidominance)
soSize :: SO -> Int
soSize (S so) = foldr (\x y -> (soSize x) + y) 1 (Set.toList so)
soSize (O (Ps ts)) = foldr (\x y -> (soSize (O x)) + y) 1 ts
soSize _ = 1

-- given LSOs, return largest, i.e. the head
maxx :: [LSO] -> LSO
maxx lsos = foldr1 (\x y ->if ((soSize.fst) x) >= ((soSize.fst) y) then x else y) lsos

-- given WS, return matching ([head,comp], [other LSOs])
match:: [WS] -> ([LSO],[LSO])
match (ws0:wss) = case List.partition ((/= []).fst.snd) (Set.toList ws0) of
    ([h],others) -> let f = (head.fst.snd) h in case (List.partition ((== f).head.snd.snd) others, wss) of
      (([c],others'),[]) -> ([h,c],others')                                            -- IM
      (([],_),[ws1]) -> case List.partition ((== f).head.snd.snd) (Set.toList ws1) of  -- EM
        ([c],others') -> ([h,c], others ++ others')

-- if LSOs satisfy shortest move constraint, return WS
smc :: [LSO] -> WS
smc lsos = if smc' [] lsos then Set.fromList lsos else error "smc violation"
  where
    smc' _ [] = True
    smc' sofar ((s,([],p:ps)):lsos) = if elem p sofar then False else smc' (p:sofar) lsos
    smc' sofar (_:lsos) = smc' sofar lsos

-- the derivational step: binary merge and label
d :: [WS] -> WS
d wss = let (matched,movers) = match wss in
        let (sos,labels) = unzip matched in
           smc (t (zip (mrg sos:tail sos) (ck labels) ++ movers))
